### DISCUSSION TITLE
2024 Cohort 2 - Group 12 Discussion

### DISCUSSION THREAD

**Discussion by sharon-digipres** [Digital-Preservation-Coalition/PythonStudyGroups#38] [General]
This is the discussion thread for those in **Group 12** of the July-December 2024 Cohort

**Comment by Karen-NMRN** [comment]
Hi, so first step accomplished and created a GitHub account! Good to meet you all earlier.

**Comment by DeanAyotte** [comment]
Hi all, I've attached the brief notes we discussed today with links etc to resources people mentioned. Nice to meet you all, and see you next month.
[DPC Python Study Group Meeting Notes 17.7.24.docx](https://github.com/user-attachments/files/16267900/DPC.Python.Study.Group.Meeting.Notes.17.7.24.docx)

**Comment by DeanAyotte** [comment]
Hey all, I've emailed the DPC about possible scripts we could work on over the next few months but I also found this example from the first cohort in their folders: https://github.com/Digital-Preservation-Coalition/PythonStudyGroups/commit/c93d63346d3102620fcb52c3c39b770efb9af6e4

**Comment by StephCornwell** [comment]
Instructions for command line python install for future reference: [python for archivists](https://ecommons.cornell.edu/items/3ad5852b-2560-41e0-95a5-ab47c25356c1), pg. 7

**Reply by hdeakyne** [reply]
Thanks for this help! Note: Not sure this works if you don't have administrator privileges. That's where I got stuck. I'm going to wait until I really need it to bother IT. Maybe I'll see if I can get it to work at home.

**Reply by hdeakyne** [reply]
Update: When I was installing at home (Windows), I noticed the install screen had a box to check to add to PATH. So I just reinstalled at work with that checked off. Much easier! Doesn't require admin.

**Comment by hdeakyne** [comment]
In case you had this question while looking at sample scripts, "How do I choose the right module?",  here is a pretty well structured explanation of the modules: https://pymotw.com/3/index.html. It was the clearest one I saw while trying to figure out what "import os", etc meant.

**Reply by hdeakyne** [reply]
Also, my department is doing group ArchivesSpace API learning. The trainer in the workshop video we are watching listed the modules/libraries she uses the most for archival work: JSON, CSV (for working with spreadsheets), and Request (for handling the request/response cycle). 

Others she mentioned that she uses, but less frequently are pymarc (for MARC21), sqlalchemy (working with SQL in ASpace), lxml, and ArchivesSnake which is ASpace specific.

**Comment by DeanAyotte** [comment]
Hey all, been using this resource recently which has been useful for some basics: https://www.codecademy.com/catalog/language/python

**Comment by StephCornwell** [comment]
I've been using codedex, nice easy to follow basic lessons and I enjoy the fun aesthetic: https://www.codedex.io/python

**Reply by DeanAyotte** [reply]
This one is great, thanks for the link!

**Reply by hdeakyne** [reply]
Ah, it's so cute!

**Comment by hdeakyne** [comment]
I went through the Codedex (so fun!) until I got to the paid lessons. Here is a little script I made afterwards; most of it came from Googling how to do the things I wanted to do. I'm traveling soon and wanted to make a list of outfits to wear from the clothing I plan to bring. It runs in the Codedex screen, but can't get it to work right in the command line or Python IDLE on my computer. 
[outfits.txt](https://github.com/user-attachments/files/16639293/outfits.txt)

Also, maybe I can add on a loop or something so that it will pick 6 outfits. We'll see!

**Reply by hdeakyne** [reply]
Yes, here it is with six different outfits listed. I tried to use the range function to number them, but couldn't figure that one out. How to get a number in front of each outfit. Somehow using print(f'{i}  ???
[outfits_updated.txt](https://github.com/user-attachments/files/16639526/outfits_updated.txt)

**Reply by StephCornwell** [reply]
Had a play and I think this does what you are after: 
print ('Day ' + str(outfits+1) + ': ' + selected_top, selected_layer, selected_pant, selected_shoes, sep=", ")

[If you add the +1 to outfits it will print beginning at 1 rather than 0]

You can also just print it as follows if not bothered about the punctuation or additional text before number
print (outfits+1, selected_top, selected_layer, selected_pant, selected_shoes, sep=", ")

I think the f{ } only works if you are working inside a for loop.

**Reply by StephCornwell** [reply]
Wrong about the f{ }! Here's the same thing expressed without having to use as much syntax: 

print (f"Day {outfits+1}: {selected_top}, {selected_layer}, {selected_pant}, {selected_shoes}")

https://www.geeksforgeeks.org/formatted-string-literals-f-strings-python/

**Reply by hdeakyne** [reply]
Yes, amazing, thank you! I tried so many different ways to getting the f{} to work, but couldn't find the right syntax.

**Reply by hdeakyne** [reply]
P.S. just to see if it would work, I made the script even smaller. 
[outfits2.txt](https://github.com/user-attachments/files/16692446/outfits2.txt)

**Reply by StephCornwell** [reply]
Nice!

[a comment by an automated account is not shown]

**Comment by harrylloyd-bl** [comment]
Hi Group 12 if you need anything @ me or use the [questions](https://github.com/Digital-Preservation-Coalition/PythonStudyGroups/discussions/categories/questions) section 😄

**Comment by hdeakyne** [comment]
Related to our possible group project which involves referring to CSV files to do things, the API video I shared discusses at the end (3:26) writing script that pulls info from CSV to create new records in ASpace. 
![CSV_to_JSON](https://github.com/user-attachments/assets/f49f28a4-4ab7-4e65-8eed-362c746952b5)

**Comment by DeanAyotte** [comment]
I've had a go at the file renamer we talked about trying out in our last meeting: https://github.com/Digital-Preservation-Coalition/PythonStudyGroups/blob/main/How-Tos/Use%20Case%20-%20How%20to%20batch%20rename%20files.md?plain=1

Attached is my version of the script with some slight variations from the code that is suggested in the step by step guide.

I thought maybe something that might be useful in the file renamer script is to have an output file which acts as a log without having manually export text file from Terminal or Command Line using manual inputs. I've been trying to figure out a way to do this as part of the python script into a specific directory folder but I can't find a straight answer online to the issues I'm running into. Anyone got any suggestions?

[renamer.txt](https://github.com/user-attachments/files/16763078/renamer.txt)

**Reply by hdeakyne** [reply]
Since so many people talked about using ChatGPT to get Python solutions, I tried that (Gemini) and it gave me this. I haven't had a chance to test it yet! And it probably doesn't work since AI. But basically it looks like we need to use scripts in the CSV library/module to create a list showing old_names and new_names? 
[AI_fileOutputSuggestions.txt](https://github.com/user-attachments/files/16764103/AI_fileOutputSuggestions.txt)

Would we create a file with the list of old_names already there, add a column with the title field called new_names, and ask it to print the new names to this column in that file?

**Reply by DeanAyotte** [reply]
I've tested this a bit this morning and it generates a csv with a header for old filename and new filename but with no other information. I'll see if I can find anything more about generating some kind of change log. Possibly @harrylloyd-bl might know how to do this?

**Reply by harrylloyd-bl** [reply]
Morning @DeanAyotte, the Gemini script should work alright to produce you a csv with your outputs. For logging in Python you can use the inbuilt [`logging`](https://docs.python.org/3/howto/logging.html#logging-basic-tutorial) package. `logging` has the added benefit that most open source packages also use `logging`, so in more complex scripts info from the activity of other packages will be logged without you having to do anything, which is great for debugging!

**Reply by DeanAyotte** [reply]
Ok, I'm not sure what it is going on then. I've attached the terminal output, csv file that is generated and the code in full if you have any guidance. Thanks Harry!

[file_renamer.txt](https://github.com/user-attachments/files/16781598/file_renamer.txt)
[renamed_files.csv](https://github.com/user-attachments/files/16781599/renamed_files.csv)
[Terminal_Saved_Output.txt](https://github.com/user-attachments/files/16781608/Terminal_Saved_Output.txt)

**Reply by harrylloyd-bl** [reply]
Great work so far, though as you've found there's a couple problems with the script at the moment. 

First, you have your code (up to line 23) which renames files and prints a message to the console, then you have the Gemini part of the script which renames files and saves renamed file names into a csv. As you're repeating the renaming logic, by the time the Gemini code runs all files have already been renamed, so when it looks for files to rename it doesn't find any and doesn't write anything to the csv. You could either adopt the csv export part of the Gemini code into your code, or move the writing to console part of your code into the Gemini code. 

Second, for the Gemini part of the script to work you'd need to assign old_pattern and new_pattern (lines 59/60) to the patterns you're trying to replace, I'll leave it to you to work out what they'd need to be based on your comment in line 11!

**Reply by DeanAyotte** [reply]
Changed some things based on your suggestions and then ran it through Chat GPT4 for a final check and amended it further. This works now as a file renamer which generates a csv to track changes.

Thanks for the help @harrylloyd-bl and @hdeakyne!

[file_renamer.txt](https://github.com/user-attachments/files/16785789/file_renamer.txt)

I'll add this code to the Repo for our group once I have write access.

**Reply by hdeakyne** [reply]
Cool!  I'm going on vacation soon so don't have time to work on Python for the next few weeks :( But I'm very excited to look at this all when I get back.

**Reply by hdeakyne** [reply]
Yay, I finally had a chance to try it out. I had it search for "IMG" and change it to another word to better describe the images. It only changed images ending in .JPG (I updated that too, I think case might matter??) and skipped over all the .AAE and .MOV files.

Initially when I tried it out, I got an error (pictured). Looking it up meant I needed to add an r in front of the source path to produce a raw string: https://stackoverflow.com/questions/1347791/unicode-error-unicodeescape-codec-cant-decode-bytes-when-writing-windows. I'm just quoting from this link, I don't actually fully understand the reasoning.
![Error](https://github.com/user-attachments/assets/53043c8c-0e99-48ed-875b-24824547b1f8)

**Reply by hdeakyne** [reply]
Now I'm curious if there is a way to get it to search within folders within the source folder. Or can we add multiple source paths?

**Reply by DeanAyotte** [reply]
Interesting. I'm wondering if the source folder error is a difference between Mac (which I am using at home) and Windows.

It would be good if multiple source folder paths could be opened at the same time. WIth this code I imagine you'd have to pull all the files into a single ingest/working folder which introudces possible errors and extra time.

**Reply by hdeakyne** [reply]
Yes, I now see that the How To discusses the other solution for Mac to Windows, using double backslashes. 
I've tried a few solutions for doing multiple folders at once, but nothing yet.

**Comment by DeanAyotte** [comment]
Here is a list of all the functions that Python can do: https://docs.python.org/3/library/functions.html

**Comment by hdeakyne** [comment]
Meeting 3, September 18, 2024, Notes. The meeting was just Dean and me! So we didn't really move forward much. Decided to continue with the decisions made in Meeting 2 on what to work on (working through various tutorials) and see if we can come up with a way to have the renaming program work on multiple folders at once.

**Comment by DeanAyotte** [comment]
Hey all,

I've been learning about reading/writing files and creating csv files in a python script and I have tried to apply this to extracting and converting data in an xml file into a csv spreadsheet. 

I am using metadata from the National Library of Scotland's Moving Image Archive that is available on their Data Foundry here: https://data.nls.uk/data/metadata-collections/moving-image-archive/

The script I have written to do this is attached. It currently creates a csv file with the different DC elements as row headers, but does not fill in any of the data from the xml file.

I've tried running it through Chat GPT 4 for some assistance but can't work out what is happening. Any ideas?

Thanks!

[data_extractor.txt](https://github.com/user-attachments/files/17204698/data_extractor.txt)

**Reply by DeanAyotte** [reply]
I've had a bit more of a test and this version seems to work. Open to more suggestions and tips though 😃 

[data_extractor_v2.txt](https://github.com/user-attachments/files/17208610/data_extractor_v2.txt)

**Reply by harrylloyd-bl** [reply]
Looks like you got the answer! To do with setting the right namespaces and doing the right `findall()`s? 

For debugging generally I can really recommend Julia Evans' [poster](https://social.jvns.ca/@b0rk/112955439024370364) and [zine](https://wizardzines.com/zines/debugging-guide/) (if your budget can spare $12) on debugging as a good framework to follow. I do pretty much everything in there any time I find a bug. Specific Python debuggers vary but the VS Code/Spyder/Pycharm/Jupyter notebook ones are good and the interactive Python shell or a notebook are also helpful tools!

**Reply by DeanAyotte** [reply]
Thanks @harrylloyd-bl, I'll grab those next pay day. I've been using VS Code for a bit, sometimes been finding the debugger a bit confusing, but that'll probably come with experience.

**Comment by DeanAyotte** [comment]
Hey all,

I was having a look at using python to try and write a script to extract metadata from AV files and I ended up finding this information with lots of scripts for extracting metadata from different file formats: https://www.linkedin.com/pulse/mastering-metadata-extracting-file-information-python-yamil-garcia-wurwe

I'm going to try running and tweaking the AV one, so I'll let you know if there are any issues/solutions.

**Comment by harrylloyd-bl** [comment]
Hi group 12, sorry to not drop in to your session today, if there's anything you need help with let us know :)
